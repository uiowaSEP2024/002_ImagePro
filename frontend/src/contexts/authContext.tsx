import { fetchCheckUserLoggedIn, fetchLogin, fetchLogout } from "@/data";
import { User } from "@/data/types";
import { useRouter } from "next/router";
import React, { createContext, useEffect, useState } from "react";

interface IAuthContextProps {
  currentUser?: User;
  isAuthenticating: boolean;
  setCurrentUser: (user: any) => void;
  setIsAuthenticating: (loading: boolean) => void;
  logOut: () => Promise<void>;
  logIn: (email: string, password: string) =>  Promise<{user: User}|void>;
  refreshAuth: () => Promise<void>;
}

export const AuthContext = createContext<IAuthContextProps>({
  currentUser: undefined,
  isAuthenticating: true,
  setCurrentUser: () => {},
  setIsAuthenticating: () => {},
  logOut: async () =>{},
  logIn: async () =>{},
  refreshAuth: async () =>{}
});

export const AuthContextProvider = (props: React.PropsWithChildren) => {
  const [currentUser, setCurrentUser] = useState<User>();
  const [isAuthenticating, setIsAuthenticating] = useState(true);
  const router = useRouter();



  useEffect(()=>{
    refreshAuth()
  }, [])

  const logOut = async () => {
    try {
      await fetchLogout();
      await refreshAuth()
      router.push("/login");
    } catch (e) {
      throw e
    }
  };


  const logIn = async (email:string, password: string) => {
    let data
    try {
      data = await fetchLogin(email, password);
      await refreshAuth()
    } catch (e) {
      throw e
    }
    return data
  };

  const refreshAuth = async () =>{
    try{
      const data = await fetchCheckUserLoggedIn()
      setCurrentUser(data.user)
    }catch(e){
      setCurrentUser(undefined)
    }
    setIsAuthenticating(false)
  }


  return (
    <AuthContext.Provider
      value={{
        currentUser,
        isAuthenticating,
        setCurrentUser,
        setIsAuthenticating,
        logOut,
        logIn,
        refreshAuth
      }}
    >
      {props.children}
    </AuthContext.Provider>
  );
};
