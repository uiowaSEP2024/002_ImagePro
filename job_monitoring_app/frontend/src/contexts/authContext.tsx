/**
 * This file contains the AuthContext and AuthContextProvider components of the application.
 * The AuthContext is a React context that provides a way to pass the authentication data down the component tree without having to pass props down manually at every level.
 * The AuthContextProvider is a React component that provides the actual data and functions that will be passed down to the components that subscribe to the AuthContext.
 * It fetches the current user data from the server, and provides functions for logging in, logging out, and refreshing the authentication data.
 *
 * https://react.dev/reference/react/createContext
 * https://vercel.com/guides/react-context-state-management-nextjs
 *
 */

// Import necessary libraries, components, hooks, and types.
import { fetchCheckUserLoggedIn, fetchLogin, fetchLogout } from "@/data";
import { User } from "@/data/types";
import { useRouter } from "next/router";
import React, { createContext, useEffect, useState } from "react";

// Define the type for the AuthContext props.
interface IAuthContextProps {
  currentUser?: User;
  isAuthenticating: boolean;
  setCurrentUser: (user: any) => void;
  setIsAuthenticating: (loading: boolean) => void;
  logOut: () => Promise<void>;
  logIn: (email: string, password: string) =>  Promise<{user: User}|void>;
  refreshAuth: () => Promise<void>;
}

/**
 * The AuthContext is a React context that provides a way to pass the authentication data down the component tree without having to pass props down manually at every level.
 */
export const AuthContext = createContext<IAuthContextProps>({
  currentUser: undefined,
  isAuthenticating: true,
  setCurrentUser: () => {},
  setIsAuthenticating: () => {},
  logOut: async () =>{},
  logIn: async () =>{},
  refreshAuth: async () =>{}
});

/**
 * The AuthContextProvider is a React component that provides the actual data and functions that will be passed down to the components that subscribe to the AuthContext.
 * It fetches the current user data from the server, and provides functions for logging in, logging out, and refreshing the authentication data.
 */
export const AuthContextProvider = (props: React.PropsWithChildren) => {
  // Initialize state variables for the current user data and the authentication loading state.
  const [currentUser, setCurrentUser] = useState<User>();
  const [isAuthenticating, setIsAuthenticating] = useState(true);
  const router = useRouter();

  // Refresh the authentication data when the component mounts.
  useEffect(()=>{
    refreshAuth()
  }, [])

  // Define the logOut function.
  const logOut = async () => {
    try {
      await fetchLogout();
      await refreshAuth()
      router.push("/login");
    } catch (e) {
      throw e
    }
  };

  // Define the logIn function.
  const logIn = async (email:string, password: string) => {
    let data
    try {
      data = await fetchLogin(email, password);
      await refreshAuth()
    } catch (e) {
      throw e
    }
    return data
  };

  // Define the refreshAuth function.
  const refreshAuth = async () =>{
    try{
      const data = await fetchCheckUserLoggedIn()
      setCurrentUser(data.user)
    }catch(e){
      setCurrentUser(undefined)
    }
    setIsAuthenticating(false)
  }

  // Render the AuthContextProvider component.
  return (
    <AuthContext.Provider
      value={{
        currentUser,
        isAuthenticating,
        setCurrentUser,
        setIsAuthenticating,
        logOut,
        logIn,
        refreshAuth
      }}
    >
      {props.children}
    </AuthContext.Provider>
  );
};
