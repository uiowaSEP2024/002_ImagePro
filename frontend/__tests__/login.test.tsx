import { act, render, screen, waitFor } from "@testing-library/react";
import Login from "@/pages/login";
import "@testing-library/jest-dom";
import { useRouter } from "next/router";
import { AuthContextProvider } from "@/contexts/authContext";

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/",
      pathname: "",
      query: "",
      asPath: "",
      push: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn()
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null)
    };
  }
}));

jest.mock("@/data", () => ({
  fetchCheckUserLoggedIn() {
    return new Promise((resolve) => {
      resolve({ detail: "Not authenticated" });
    });
  }
}));

describe("Login", () => {
  it("renders text", async () => {
    await act(async () => render(<Login />, { wrapper: AuthContextProvider }));

    expect(useRouter().push).not.toBeCalledWith("/");

    const text = await waitFor(() =>
      screen.getByRole("heading", {
        name: /Login/i
      })
    );

    expect(text).toBeInTheDocument();
  });
});
