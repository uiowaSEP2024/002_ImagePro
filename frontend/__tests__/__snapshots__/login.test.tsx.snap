// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders Login unchanged 1`] = `
<div>
  <div>
    <div
      class="nextui-c-fdHeMm nextui-c-fdHeMm-dwxLNB-responsive-true nextui-c-fdHeMm-iblJuSr-css"
    >
      <div
        class="nextui-c-BDLTQ nextui-c-jMTBsW nextui-c-gulvcB nextui-c-BDLTQ-kUADSW-variant-bordered nextui-c-BDLTQ-fmlsco-borderWeight-light nextui-c-BDLTQ-cuwTXc-disableAnimation-false nextui-c-BDLTQ-ilimyxf-css"
        role="section"
        tabindex="-1"
      >
        <p
          class="nextui-c-PJLV nextui-c-PJLV-hyvuql-weight-bold nextui-c-PJLV-ibsXxfh-css"
        >
          Login
        </p>
        <div
          class="nextui-c-iwmZdB nextui-c-iwmZdB-jTlWbp-color-primary nextui-c-iwmZdB-jTfqsF-helperColor-default nextui-c-iwmZdB-dpkacN-borderWeight-normal nextui-c-iwmZdB-bkiCag-size-lg nextui-c-iwmZdB-ifGHEql-css nextui-input-main-container nextui-input-main-container--normal"
          data-state="normal"
        >
          <div
            class="nextui-c-boZsAS nextui-c-boZsAS-fjttfO-isTextarea-false nextui-c-boZsAS-iNKybL-animated-true nextui-c-boZsAS-AmGUN-cv nextui-input-container nextui-input-container--normal nextui-input-container--input"
          >
            <label
              class="nextui-c-eXOOPO nextui-c-eXOOPO-iNKybL-shadow-true nextui-c-eXOOPO-gWufxv-underlined-true nextui-c-eXOOPO-lkBNdM-isTextarea-false nextui-input-wrapper nextui-input-wrapper--normal nextui-input-wrapper--underlined nextui-input-wrapper--shadow"
            >
              <span
                class="nextui-c-PJLV nextui-c-iFHsrU nextui-input-placeholder"
              >
                Email
              </span>
              <input
                aria-label="Email"
                aria-multiline="false"
                aria-placeholder="Email"
                aria-readonly="false"
                aria-required="false"
                autocomplete="off"
                class="nextui-c-jeuecp nextui-input"
                data-state="normal"
                id="react-aria8428947043-1"
                placeholder="Email"
                type="text"
                value=""
              />
              <button
                aria-label="clear"
                class="nextui-c-ctQnhE nextui-c-ctQnhE-kClpVX-underlined-true nextui-input-clear-button"
                lang="en"
                type="button"
              >
                <span
                  class="nextui-c-gDSmsE nextui-c-gulvcB"
                >
                  <svg
                    class=""
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                  >
                    <path
                      d="M12 2a10 10 0 1010 10A10.016 10.016 0 0012 2zm3.36 12.3a.754.754 0 010 1.06.748.748 0 01-1.06 0l-2.3-2.3-2.3 2.3a.748.748 0 01-1.06 0 .754.754 0 010-1.06l2.3-2.3-2.3-2.3A.75.75 0 019.7 8.64l2.3 2.3 2.3-2.3a.75.75 0 011.06 1.06l-2.3 2.3z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </button>
            </label>
          </div>
          <div
            class="nextui-c-gJvMas nextui-input-helper-text-container"
          />
        </div>
        <span
          aria-hidden="true"
          class="nextui-c-gNVTSf nextui-c-gNVTSf-hakyQ-inline-false nextui-c-gNVTSf-igvLppd-css"
        />
        <div
          class="nextui-c-iwmZdB nextui-c-iwmZdB-jTlWbp-color-primary nextui-c-iwmZdB-jTfqsF-helperColor-default nextui-c-iwmZdB-dpkacN-borderWeight-normal nextui-c-iwmZdB-bkiCag-size-lg nextui-c-iwmZdB-idXLuPy-css nextui-input-main-container nextui-input-main-container--normal"
          data-state="normal"
        >
          <div
            class="nextui-c-boZsAS nextui-c-boZsAS-fjttfO-isTextarea-false nextui-c-boZsAS-iNKybL-animated-true nextui-c-boZsAS-AmGUN-cv nextui-input-container nextui-input-container--normal nextui-input-container--input"
          >
            <label
              class="nextui-c-eXOOPO nextui-c-eXOOPO-iNKybL-shadow-true nextui-c-eXOOPO-gWufxv-underlined-true nextui-c-eXOOPO-lkBNdM-isTextarea-false nextui-input-wrapper nextui-input-wrapper--normal nextui-input-wrapper--underlined nextui-input-wrapper--shadow"
            >
              <span
                class="nextui-c-PJLV nextui-c-iFHsrU nextui-input-placeholder"
              >
                Password
              </span>
              <input
                aria-label="Password"
                aria-multiline="false"
                aria-placeholder="Password"
                aria-readonly="false"
                aria-required="false"
                autocomplete="off"
                class="nextui-c-jeuecp nextui-input"
                data-state="normal"
                id="react-aria8428947043-3"
                placeholder="Password"
                type="text"
                value=""
              />
              <button
                aria-label="clear"
                class="nextui-c-ctQnhE nextui-c-ctQnhE-kClpVX-underlined-true nextui-input-clear-button"
                lang="en"
                type="button"
              >
                <span
                  class="nextui-c-gDSmsE nextui-c-gulvcB"
                >
                  <svg
                    class=""
                    height="24"
                    viewBox="0 0 24 24"
                    width="24"
                  >
                    <path
                      d="M12 2a10 10 0 1010 10A10.016 10.016 0 0012 2zm3.36 12.3a.754.754 0 010 1.06.748.748 0 01-1.06 0l-2.3-2.3-2.3 2.3a.748.748 0 01-1.06 0 .754.754 0 010-1.06l2.3-2.3-2.3-2.3A.75.75 0 019.7 8.64l2.3 2.3 2.3-2.3a.75.75 0 011.06 1.06l-2.3 2.3z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </button>
            </label>
          </div>
          <div
            class="nextui-c-gJvMas nextui-input-helper-text-container"
          />
        </div>
        <div
          class="nextui-c-dfbgCO nextui-c-dfbgCO-fGHEql-fluid-true nextui-c-dfbgCO-igmUkpg-css"
        >
          <a
            class="nextui-c-kzikkm nextui-c-gulvcB nextui-c-kzikkm-iVlgeG-color-secondary nextui-c-kzikkm-fZgbCn-block-true nextui-c-kzikkm-isWIbj-animated-true nextui-c-kzikkm-jSBLRj-cv nextui-c-kzikkm-iclraoi-css nextui-link"
            href="/signup"
          >
            New user? Create Account. 
          </a>
        </div>
        <span
          aria-hidden="true"
          class="nextui-c-gNVTSf nextui-c-gNVTSf-hakyQ-inline-false nextui-c-gNVTSf-igvLppd-css"
        />
        <button
          class="nextui-c-iWjDFM nextui-c-gulvcB nextui-c-iWjDFM-hkKLfu-color-default nextui-c-iWjDFM-gkIlwi-size-md nextui-c-iWjDFM-cwXrJp-borderWeight-normal nextui-c-iWjDFM-iPJLV-css nextui-button nextui-button--ready"
          data-state="ready"
          type="button"
        >
          <span
            class="nextui-button-text"
          >
            Log in
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
